tracer/src/Datadog.Trace/Debugger/Conditions/ProbeConditionExpressionParser.cs

```diff
@@ -0,0 +1,664 @@
+            }
+
+            return MemberPathExpression(parts.Skip(1).ToArray(), member);
+        }
+
+        private Expression MemberPathExpression(string[] parts, ParameterExpression expression)
+        {
+            try
+            {
+                var member = Expression.Field(expression, parts[0]);
+                foreach (var currentField in parts.Skip(1))
+                {
+                    member = Expression.Field(member, currentField);
+                }
+
+                return member;
+            }
+            catch
+            {
+                return Expression.Constant(UndefinedValue.Instance);
+            }
+        }
+
+        private MethodInfo GetMethodByReflection(Type type, string name, Type[] parametersTypes)
+        {
+            // todo: cache method in a static field
+            const BindingFlags bindingFlags = BindingFlags.Static | BindingFlags.Instance | BindingFlags.InvokeMethod |
+                                              BindingFlags.NonPublic | BindingFlags.Public;
+            var method = type.GetMethod(name, bindingFlags, null, parametersTypes, null);
+            if (method == null)
+            {
+                throw new NullReferenceException($"{type.FullName}.{name} method not found");
+            }
+
+            return method;
+        }
+
+        private MethodCallExpression CollectionCountExpression(Expression source)
+        {
+            if (!IsTypeSupportCount(source))
+            {
+                throw new InvalidOperationException("Source must be an array or implement ICollection or IReadOnlyCollection");
+            }
+
+            var countOrLength = GetMethodByReflection(source.Type, source.Type.IsArray ? "get_Length" : "get_Count", Type.EmptyTypes);
+
+            return Expression.Call(source, countOrLength);
+        }
+
+        private bool IsSafeCollection(Expression source)
+        {
+            return source.Type.GetInterface("ICollection") != null ||
+                   source.Type.GetInterface("IReadOnlyCollection") != null ||
+                   source.Type.IsArray;
+        }
+
+        private bool IsCollection(Expression source)
+        {
+            return IsSafeCollection(source) || source.Type.GetInterface("IEnumerable") != null;
+        }
+
+        private bool IsTypeSupportCount(Expression source)
+        {
+            return source.Type.GetInterface("ICollection") != null ||
+                   source.Type.GetInterface("IReadOnlyCollection") != null ||
+                   source.Type.IsArray;
+        }
+
+        private bool IsTypeSupportIndex(Expression source)
+        {
+            return source.Type.GetInterface("IList") != null || source.Type.GetInterface("IReadOnlyList") != null;
+        }
+
+        private Tuple<Expression<Func<ScopeMember, ReadOnlyCollection<ScopeMember>, bool>>, string> ParseExpression(string conditionJson, ScopeMember @this, ReadOnlyCollection<ScopeMember> argsOrLocals)
```

Comment:
The Tuple that is being returned here is pretty complicated and makes the code kind of hard to follow - considering extract a class/record for it?